File: 130-surrounded-regions.java
/*
 Given an m x n matrix board containing 'X' and 'O', capture all regions that are 4-directionally surrounded by 'X'.

A region is captured by flipping all 'O's into 'X's in that surrounded region.

*/
class Solution {
    public void solve(char[][] board) {
        if(board.length == 0 || board[0].length == 0) return;

        int m = board.length;
        int n = board[0].length;

        for(int i = 0; i < m; i++) {
            capture(board, i, 0);
            capture(board, i, n - 1);
        }
        
        for(int j = 1; j < n - 1; j++) {
            capture(board, 0, j);
            capture(board, m - 1, j);
        }

        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
               if( board[i][j] == 'O') board[i][j] = 'X';
                                if(board[i][j] == 'T') board[i][j] = 'O';
            }
        }

    }

    private void capture(char[][] board, int i, int j) {
        if(i < 0 || i >= board.length || j < 0 || j >= board[0].length) return;
        if(board[i][j] == 'X' || board[i][j] == 'T') return;
        board[i][j] = 'T';
        capture(board, i - 1, j);
        capture(board, i + 1, j);
        capture(board, i, j - 1);
        capture(board, i, j + 1);
    }
}

File: Javascript/130-surrounded-regions.js
/*
 Given an m x n matrix board containing 'X' and 'O', capture all regions that are 4-directionally surrounded by 'X'.

A region is captured by flipping all 'O's into 'X's in that surrounded region.

*/

var solve = function(board) {
    for (let r = 0; r < board.length; r++) {
        for (let c = 0; c < board[r].length; c++) {
            if (board[r][c] === "O" && isOnBorder(r, c)) {
                checkNeighbours(r, c);
            }
        }
    }

    function isOnBorder(r, c) {
        return r === 0 || r === board.length - 1 || c === 0 || c === board[r].length - 1;
    }

    function checkNeighbours(r, c) {
        if (isBeyondBoard(r, c) || board[r][c] !== "O") {
            return;
        }

        board[r][c] = "Visited";

        checkNeighbours(r + 1, c);
        checkNeighbours(r, c + 1);
        checkNeighbours(r - 1, c);
        checkNeighbours(r, c - 1);
    }

    function isBeyondBoard(r, c) {
        return r < 0 || r >= board.length || c < 0 || c >= board[r].length;
    }

    for (let r = 0; r < board.length; r++) {
        for (let c = 0; c < board[r].length; c++) {
            if (board[r][c] === "Visited") {
                board[r][c] = "O"
            } else {
                board[r][c] = "X"
            }
        }
    }
};
